[Ingest pipelines] Create pipeline UI

alisonelizabeth opened the pull request:
This PR adds the create pipeline page to the ingest pipelines app.

### Screenshots 
![create_pipeline](https://user-images.githubusercontent.com/5226211/79085014-d5e33580-7d04-11ea-8561-384c66459f5e.png)

Optional fields enabled:

![create_pipeline_2](https://user-images.githubusercontent.com/5226211/79085023-ded40700-7d04-11ea-9616-c7e8de5cfdec.png)

Field validation:
![create_pipeline_validation](https://user-images.githubusercontent.com/5226211/79085106-3c685380-7d05-11ea-81a9-ad718c645b7b.png)

elasticmachine commented:
Pinging @elastic/es-ui (Team:Elasticsearch UI)

kibanamachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/39724/)
* Commit: 90ec298b0ecc1734774a6a656171e49fc32606ff
* [Pipeline Steps](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/39724/flowGraphTable) (look for red circles / failed steps)
* [Interpreting CI Failures](https://www.elastic.co/guide/en/kibana/current/interpreting-ci-failures.html)

### Failed CI Steps
* [Execute kibana-intake](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/39724/execution/node/151/log)

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/39724/","number":"39724","commit":"90ec298b0ecc1734774a6a656171e49fc32606ff"}]}
PIPELINE-->

kibanamachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/39958/)
* Commit: 1e4a8d37343bc23ec0afd7cbe8c08313d61c58e9

### History
* :broken_heart: [Build #39724](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/39724/) failed 90ec298b0ecc1734774a6a656171e49fc32606ff

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/39724/","number":"39724","commit":"90ec298b0ecc1734774a6a656171e49fc32606ff"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/39958/","number":"39958","commit":"1e4a8d37343bc23ec0afd7cbe8c08313d61c58e9"}]}
PIPELINE-->

kibanamachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/40208/)
* Commit: 10b9210e968b6c0bee81f0312dc025fc6ed3ca2b

### History
* :green_heart: [Build #39958](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/39958/) succeeded 1e4a8d37343bc23ec0afd7cbe8c08313d61c58e9
* :broken_heart: [Build #39724](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/39724/) failed 90ec298b0ecc1734774a6a656171e49fc32606ff

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/39724/","number":"39724","commit":"90ec298b0ecc1734774a6a656171e49fc32606ff"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/39958/","number":"39958","commit":"1e4a8d37343bc23ec0afd7cbe8c08313d61c58e9"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/40208/","number":"40208","commit":"10b9210e968b6c0bee81f0312dc025fc6ed3ca2b"}]}
PIPELINE-->

sebelga commented on x-pack/plugins/ingest_pipelines/public/application/components/pipeline_form/schema.tsx:
```diff
@@ -0,0 +1,132 @@
+    label: i18n.translate('xpack.ingestPipelines.form.processorsFieldLabel', {
+      defaultMessage: 'Processors',
+    }),
+    serializer: processors => parseJson(processors),
```
Nit: can be simply written: `serializer: parseJson`

sebelga commented on x-pack/plugins/ingest_pipelines/public/application/components/pipeline_form/schema.tsx:
```diff
@@ -0,0 +1,132 @@
+      defaultMessage: 'Processors',
+    }),
+    serializer: processors => parseJson(processors),
+    deserializer: processors => stringifyJson(processors),
```
Nit: here the same `deserializer: stringifyJson `

kibanamachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/40412/)
* Commit: 6bba3dacf16071f2ce0ac5519bfc2a38e684704f

### History
* :green_heart: [Build #40208](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/40208/) succeeded 10b9210e968b6c0bee81f0312dc025fc6ed3ca2b
* :green_heart: [Build #39958](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/39958/) succeeded 1e4a8d37343bc23ec0afd7cbe8c08313d61c58e9
* :broken_heart: [Build #39724](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/39724/) failed 90ec298b0ecc1734774a6a656171e49fc32606ff

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/39724/","number":"39724","commit":"90ec298b0ecc1734774a6a656171e49fc32606ff"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/39958/","number":"39958","commit":"1e4a8d37343bc23ec0afd7cbe8c08313d61c58e9"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/40208/","number":"40208","commit":"10b9210e968b6c0bee81f0312dc025fc6ed3ca2b"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/40412/","number":"40412","commit":"6bba3dacf16071f2ce0ac5519bfc2a38e684704f"}]}
PIPELINE-->

jloleysens commented on x-pack/plugins/ingest_pipelines/public/application/components/pipeline_form/schema.tsx:
```diff
@@ -0,0 +1,132 @@
+const stringifyJson = (json: { [key: string]: unknown }): string =>
+  Array.isArray(json) ? JSON.stringify(json, null, 2) : '[\n\n]';
+
+const parseJson = (jsonString: string) => {
```
Could we add a return type annotation here? Something like:

```typescript
const parseJson = (jsonString: string): object[] => ...
```

jloleysens commented on x-pack/plugins/ingest_pipelines/public/application/components/pipeline_form/schema.tsx:
```diff
@@ -0,0 +1,132 @@
+  Array.isArray(json) ? JSON.stringify(json, null, 2) : '[\n\n]';
+
+const parseJson = (jsonString: string) => {
+  let parsedJSON;
```
Nit: we should probably add an explicit type here, `any` I am guessing.

jloleysens commented on x-pack/plugins/ingest_pipelines/public/application/services/breadcrumbs.ts:
```diff
@@ -0,0 +1,62 @@
+
+    this.setBreadcrumbsHandler = setBreadcrumbsHandler;
+
+    this.breadcrumbs.home = [
```
These declarations do not have to happen at setup time, it looks like they are not derived from passed in state so perhaps we can declare them on line 19. Wdyt?

jloleysens replied:
Then setup would just be

```
this.setBreadcrumbsHandler = setBreadcrumbsHandler;
```

alisonelizabeth replied:
Great catch! Fixed.

sebelga commented on x-pack/plugins/ingest_pipelines/public/application/components/pipeline_form/pipeline_form.tsx:
```diff
@@ -0,0 +1,289 @@
+                  iconType="check"
+                  onClick={form.submit}
+                  data-test-subj="submitButton"
+                  disabled={form.isValid === false}
```
Here you might need `form.isSubmitted && form.isValid === false`? Haven't tested locally but this way the button does not change state multiple time meanwhile the user fills the form.

sebelga commented on x-pack/plugins/ingest_pipelines/public/application/components/pipeline_form/pipeline_form.tsx:
```diff
@@ -0,0 +1,289 @@
+        </>
+      ) : null}
+
+      <Form form={form} data-test-subj="pipelineForm">
```
Here you can add the EUI props to display error on top of the form

```js
<Form
  form={form}
  isInvalid={form.isSubmitted && !form.isValid}
  error={form.getErrors()}
  data-test-subj="pipelineForm"
>
```

cjcenizal commented:
@alisonelizabeth Re validation, I think it would make sense to tweak the UX in a couple areas. I'd suggest:

1. Allowing the user to not define any processors
2. Allowing the user to not define any failure processors

This request is valid:

```
PUT _ingest/pipeline/my-pipeline-id
{
  "description" : "describe pipeline",
  "processors" : [
  ]
}
```

If a user creates this pipeline via the API, it might be a jarring experience for the UI to tell them it's invalid. In terms of practical use cases, I don't know if people actually create empty pipelines like this but I can imagine a workflow where someone sets up a bunch of placeholders to identify various ingestion systems without, with the intention of fleshing out the processors later. WDYT?

alisonelizabeth commented:
Thanks @sebelga and @jloleysens for the review! I believe I addressed all feedback.

> We should also add the navigation to the creation page on the empty list view (main.tsx L68).

Thanks for catching this! I noticed this while working on the edit view, but forgot to push the change to this branch. The latest commit addresses this.

> Currently the form allows me to create a pipeline with no processors. Should we stop the user from performing this action?

I agree with CJ on this. The API allows you to send an empty array for processors, so I think we should follow this behavior.

The current UI behaves as follows:
- **Processors editor**: User must specify an empty array or an array of processors. A blank editor will result in a validation error.
- **Failed processors editor**: Users must specify an array of processors or can leave the editor blank. An empty array will result in a validation error (it's invalid via the ES API).

@cjcenizal does this make sense to you?

kibanamachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/40714/)
* Commit: ef25c2265b2102dcb04be7e33cabbb9735d49ba7
* [Pipeline Steps](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/40714/flowGraphTable) (look for red circles / failed steps)
* [Interpreting CI Failures](https://www.elastic.co/guide/en/kibana/current/interpreting-ci-failures.html)

### Failed CI Steps
* [Execute kibana-intake](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/40714/execution/node/147/log)

### History
* :green_heart: [Build #40412](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/40412/) succeeded 6bba3dacf16071f2ce0ac5519bfc2a38e684704f
* :green_heart: [Build #40208](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/40208/) succeeded 10b9210e968b6c0bee81f0312dc025fc6ed3ca2b
* :green_heart: [Build #39958](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/39958/) succeeded 1e4a8d37343bc23ec0afd7cbe8c08313d61c58e9
* :broken_heart: [Build #39724](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/39724/) failed 90ec298b0ecc1734774a6a656171e49fc32606ff

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/39724/","number":"39724","commit":"90ec298b0ecc1734774a6a656171e49fc32606ff"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/39958/","number":"39958","commit":"1e4a8d37343bc23ec0afd7cbe8c08313d61c58e9"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/40208/","number":"40208","commit":"10b9210e968b6c0bee81f0312dc025fc6ed3ca2b"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/40412/","number":"40412","commit":"6bba3dacf16071f2ce0ac5519bfc2a38e684704f"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/40714/","number":"40714","commit":"ef25c2265b2102dcb04be7e33cabbb9735d49ba7"}]}
PIPELINE-->

cjcenizal commented:
@alisonelizabeth I think that's logical and mirrors the API well, but I think there are a couple rough UX edges we can smooth out. Saying "At least one on-failure processor must be defined" might confuse someone who just wants to not have any on-failure processors at all. So rather than making this validation message more verbose to communicate that, I wonder if it makes sense to just infer that this is what they want, and to delete the empty array from the API request? Likewise, if processors is blank, can we insert an empty array?

Is this worth the extra work, especially considering we intend to replace much of this code in 7.9? Not sure. I would say the higher priority is to ship as feature-complete MVP as possible for 7.8, and that UX niceties like this are nice-to-haves. As project lead it's your call.

alisonelizabeth commented:
@cjcenizal That's a valid point. I think it would improve the experience. I'm going to leave as-is for now, but will plan to revisit if there's time at the end of the cycle.

kibanamachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/40741/)
* Commit: cb8185a19ebaa79eb794d7d8e6e868bb0f6f2749
* [Pipeline Steps](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/40741/flowGraphTable) (look for red circles / failed steps)
* [Interpreting CI Failures](https://www.elastic.co/guide/en/kibana/current/interpreting-ci-failures.html)

### Failed CI Steps
* [Execute x-pack-intake](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/40741/execution/node/153/log)

---

### [Test Failures](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/40741/testReport)

<details><summary>Kibana Pipeline / x-pack-intake-agent / X-Pack Jest Tests.x-pack/plugins/remote_clusters/__jest__/client_integration.<RemoteClusterList /> when there are no remote clusters should display an empty prompt</summary>

[Link to Jenkins](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/40741/testReport//junit/X-Pack%20Jest%20Tests/x-pack_plugins_remote_clusters___jest___client_integration/Kibana_Pipeline___x_pack_intake_agent____RemoteClusterList____when_there_are_no_remote_clusters_should_display_an_empty_prompt)


#### Standard Out
```
Failed Tests Reporter:
  - Test has not failed recently on tracked branches


```

#### Stack Trace
```
Error: expect(received).toBe(expected) // Object.is equality

Expected: true
Received: false
    at Object.test (/var/lib/jenkins/workspace/elastic+kibana+pipeline-pull-request/kibana/x-pack/plugins/remote_clusters/__jest__/client_integration/remote_clusters_list.test.js:65:54)
    at Object.asyncJestTest (/var/lib/jenkins/workspace/elastic+kibana+pipeline-pull-request/kibana/node_modules/jest-jasmine2/build/jasmineAsyncInstall.js:102:37)
    at resolve (/var/lib/jenkins/workspace/elastic+kibana+pipeline-pull-request/kibana/node_modules/jest-jasmine2/build/queueRunner.js:43:12)
    at new Promise (<anonymous>)
    at mapper (/var/lib/jenkins/workspace/elastic+kibana+pipeline-pull-request/kibana/node_modules/jest-jasmine2/build/queueRunner.js:26:19)
    at promise.then (/var/lib/jenkins/workspace/elastic+kibana+pipeline-pull-request/kibana/node_modules/jest-jasmine2/build/queueRunner.js:73:41)
```
</details>

---

### History
* :broken_heart: [Build #40714](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/40714/) failed ef25c2265b2102dcb04be7e33cabbb9735d49ba7
* :green_heart: [Build #40412](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/40412/) succeeded 6bba3dacf16071f2ce0ac5519bfc2a38e684704f
* :green_heart: [Build #40208](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/40208/) succeeded 10b9210e968b6c0bee81f0312dc025fc6ed3ca2b
* :green_heart: [Build #39958](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/39958/) succeeded 1e4a8d37343bc23ec0afd7cbe8c08313d61c58e9
* :broken_heart: [Build #39724](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/39724/) failed 90ec298b0ecc1734774a6a656171e49fc32606ff

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/39724/","number":"39724","commit":"90ec298b0ecc1734774a6a656171e49fc32606ff"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/39958/","number":"39958","commit":"1e4a8d37343bc23ec0afd7cbe8c08313d61c58e9"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/40208/","number":"40208","commit":"10b9210e968b6c0bee81f0312dc025fc6ed3ca2b"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/40412/","number":"40412","commit":"6bba3dacf16071f2ce0ac5519bfc2a38e684704f"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/40714/","number":"40714","commit":"ef25c2265b2102dcb04be7e33cabbb9735d49ba7"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/40741/","number":"40741","commit":"cb8185a19ebaa79eb794d7d8e6e868bb0f6f2749"}]}
PIPELINE-->

alisonelizabeth commented:
I'm going to merge this PR into the feature branch. The latest test failure is unrelated to these changes, and the flakiness should be resolved after updating the feature branch with master.

sebelga commented:
>  I wonder if it makes sense to just infer that this is what they want, and to delete the empty array from the API request? 

Just a note, it is for those concrete use-cases that we can declare a `serializer` function at the **form** level. Adding/removing fields.
